rbd: base implementation for CSI-Addons VolumeGroup

nixpanic opened the pull request:
# Describe what this PR does #

This is the basis of the implementation for the [CSI-Addons VolumeGroup specification](https://github.com/csi-addons/spec/blob/main/volumegroup/README.md). The service and capabilities are added, but the actual implementation of the VolumeGroup creation/deletion/modifying is not included here, and will be sent as a follow-up PR.

## Is there anything that requires special attention ##

By providing this base for the CSI-Addons VolumeGroup support, reviewing should be simpler than including everything a a huge PR.

Updates: #4502

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

Madhu-1 committed:
csiaddons: register volumegroup controller

Register the volumegroup controller as part
of rbd controller server to serve the volume
group RPC spec.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

nixpanic committed:
rbd: add a ToCSI() function to the Volume interface

A VolumeGroup CSI-Addons object contains a list of CSI Volumes. A
ToCSI() function makes creating such a list much simpler.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rebase: update github.com/csi-addons/spec to latest version

Update VolumeGroup API with CreateVolumeGroupRequest that contains an
optional list of VolumeIDs.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: add Manager interface for using Volumes and VolumeGroups

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
rbd: add the VolumeGroup type

Signed-off-by: Niels de Vos <ndevos@ibm.com>

nixpanic committed:
csiaddons: initial implementation of CSI-Addons VolumeGroup

Signed-off-by: Niels de Vos <ndevos@ibm.com>

Madhu-1 commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -0,0 +1,195 @@
+	volumegroup.RegisterControllerServer(server, vs)
+}
+
+// CreateVolumeGroup RPC call to create a volume group.
+//
+// This RPC will be called by the CO to create a new volume group on behalf of
+// a user. This operation MUST be idempotent. If a volume group corresponding
+// to the specified volume group name already exists, is compatible with the
+// specified parameters in the CreateVolumeGroupRequest, the Plugin MUST reply
+// 0 OK with the corresponding CreateVolumeGroupResponse. CSI Plugins MAY
+// create the following types of volume groups:
+//
+// Create a new empty volume group or a group with specific volumes. Note that
+// N volumes with some backend label Y could be considered to be in "group Y"
+// which might not be a physical group on the storage backend. In this case, an
+// empty group can still be created by the CO to hold volumes. After the empty
+// group is created, create a new volume. CO may call
+// ModifyVolumeGroupMembership to add new volumes to the group.
```
Can we update the comment that what we are doing in cephcsi?

Madhu-1 commented on internal/csi-addons/rbd/volumegroup.go:
```diff
@@ -0,0 +1,195 @@
+//
+// Note:
+// The undocumented DO_NOT_ALLOW_VG_TO_DELETE_VOLUMES capability is set. There
+// is no need to delete eack volume that may be part of the volume group. If
```
`eack` to `each`

Madhu-1 commented on internal/rbd/manager.go:
```diff
@@ -0,0 +1,89 @@
+	"context"
+	"errors"
+
+	types "github.com/ceph/ceph-csi/internal/rbd_types"
```
can we move this to a folder inside the `internal/rbd`? like `internal/rbd/types`?

nixpanic replied:
That was not possible earlier because it would import the `internal/rbd` package, and `internal/rbd` imports the types too. It's been cleaned up quite a bit since, and maybe it is possible now.

Madhu-1 reviewed (commented):
some nits, LGTM

nixpanic committed:
rbd: move internal/rbd_types -> internal/rbd/types

The rbd_types package was initially created with references to the rbd
package. And the rbd package references the rbd_types package. Having
rbd/types was not possible due to recursive imports. After cleaning up
the rbd_types package, it can be renamed to rbd/types.

Signed-off-by: Niels de Vos <ndevos@ibm.com>

yati1998 reviewed (approved):
LGTM

Madhu-1 commented:
@mergifyio queue

nixpanic commented:
**ci/centos/mini-e2e-helm/k8s-1.28** paniced in NFS testing [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.28/detail/mini-e2e-helm_k8s-1.28/752/pipeline):
```
  [PANICKED] in [AfterEach] - /go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:261 @ 07/11/24 09:23:33.79
```

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.28

nixpanic commented:
@mergifyio requeue
